pkg/proxy/winkernel/hcnutils.go

```diff
@@ -29,15 +29,51 @@ import (
 	"strings"
 )
 
-type hnsV2 struct{}
+type HCN interface {
+	GetNetworkByName(networkName string) (*hcn.HostComputeNetwork, error)
+	ListEndpointsOfNetwork(networkId string) ([]hcn.HostComputeEndpoint, error)
+	GetEndpointByID(endpointId string) (*hcn.HostComputeEndpoint, error)
+	ListEndpoints() ([]hcn.HostComputeEndpoint, error)
+	GetEndpointByName(endpointName string) (*hcn.HostComputeEndpoint, error)
```

Comment:
This is the interface that fakeHCN and ihcn are using. The hcnutils constructor is consuming this interface to know which hcn implementation to use (the real one or the fake one). It is mainly to be able to use both hcn implementations.

Reply:
agreed we don't need the `IHCN` and `IHCNUtils`.  We should only need one interface here.  The new function can be used inside the proxier to return the correct instance, and the tests can just return the fake one that implements it.

Reply:
Looking a little closer and It look like we already have an interface (that had two implementations 
 hnsv1 and hnsv2) that does this already: https://github.com/kubernetes/kubernetes/blob/b7466d02cdceee594869b47bea7f8e1ade03b7e2/pkg/proxy/winkernel/hnsV1.go#L32-L43

We could re-use that interface and implement the fake one for the unit tests.  This would make the change set smaller and a bit more easier to understand I think

Reply:
hey, we want to mock hcn, not hcn utils, that's the reason we added the IHCN

Reply:
the current `HostNetworkService` already provides the interface to mock it.  you can do something like the following without introducing new interfaces

```
type fakehns struct {}

func (hns fakehns) getNetworkByName(name string) (*hnsNetworkInfo, error) {
   return your fake data
}
```

Reply:
The reason we fake hcn and not hns is because faking hns makes it impossible to test the hcn objects that should be created in syncproxyrules from "proxier.go". We want to test both hns and hcn, and a fake hcn implementation makes this possible.

Reply:
What I am struggling with there is that `IHCN`, `IHCNUtils` and `HostNetworkService` are all almost the same interfaces.  `HostNetworkService` has been removed and still the difference between `IHCN` and `IHCNUtils` is really minor.  There doesn't seem to be a clear distinction between the interfaces.  And after this change we have things like `https://github.dev/mpanduru/kubernetes/blob/093d57ff301711f98c3862ee44037f27be37e1f8/pkg/proxy/winkernel/hcnutils.go#L506` which still use hcn library directly.  

I see the issue was here prior to this change. The file `hnsv2` (now renamed to hcnutils) was implementing `HostNetworkService` but was using hcn directly https://github.com/kubernetes/kubernetes/blob/b7466d02cdceee594869b47bea7f8e1ade03b7e2/pkg/proxy/winkernel/hnsV2.go#L40. On top of this there were several other helper functions that did stuff. There wasn't a clear seperation prior but this doesn't help with the seperation (since there is no distinction between `IHCN` and `IHCNUtils`). Part of this might be naming of the HCN Utility and the fact that it is an addtional interface (it can just be a type here since the thing we really are mocking is the HCN library).  

This PR is doing atleast two things here: removing support for HNS v1 and refactoring the hns to add tests.  If we are dropping full support for hns1 that should be a distinct commit (and perferably a seperate PR).  After that it might be easier to tease apart the various implementations we really want here.

Reply:
I'm seeing this a bit later and things have changed since the comments above (e.g.: support for HNS v1 has been removed), but this seems fine now. Basically, ``hcnutils.go`` was untestable before simply cause of the ``hcn`` module usage, and this abstracts it, allowing this module to be tested.